src/main/java/com/minecolonies/coremod/tileentities/TileEntityBarrel.java

```diff
@@ -0,0 +1,240 @@
+package com.minecolonies.coremod.tileentities;
+
+import com.minecolonies.api.configuration.Configurations;
+import com.minecolonies.api.util.ItemStackUtils;
+import com.minecolonies.coremod.items.ModItems;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.init.Items;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.NetworkManager;
+import net.minecraft.network.play.server.SPacketUpdateTileEntity;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.EnumParticleTypes;
+import net.minecraft.util.ITickable;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TextComponentString;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldServer;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Random;
+
+public class TileEntityBarrel extends TileEntity implements ITickable
+{
+    private boolean                                 done           = false;
+    private int                                     items          = 0;
+    private int                                     timer          = 0;
+    public static final int                         MAX_ITEMS      = 64;
+    private static final int                        TIMER_END      = 24000; //24000; //Number of Minecraft ticks in 2
+                                                                            // whole days
+    private static final int                        AVERAGE_TICKS  = 20;    //The average of ticks that passes between
+                                                                            //actually ticking the tileEntity
+
+    @Override
+    public void update()
+    {
+        final World world = this.getWorld();
+
+        if (!world.isRemote && (world.getWorldTime() % (world.rand.nextInt(AVERAGE_TICKS * 2) + 1) == 0))
+        {
+            this.updateTick(world, this.getPos(), world.getBlockState(this.getPos()), new Random());
+        }
+
+    }
+
+
+    public void updateTick(final World worldIn, final BlockPos pos, final IBlockState state, final Random rand)
+    {
+        //Check if the barrel is actually full
+        if(getItems() == TileEntityBarrel.MAX_ITEMS)
+        {
+            doBarrelCompostTick(worldIn, pos, state);
+        }
+        if(this.done)
+        {
+            //If the barrel is done, we spawn particles imitating "bad smell"
+            ((WorldServer)worldIn).spawnParticle(
```

Comment:
### Errors:
- Casting worldIn to WorldServer may produce java.lang.ClassCastException

Reply:
Will never produce the exception as it never gets called if world.isRemote is true, wich means the world is a server. Not need to check if it's an instance of WorldServer and it would be too expensive.